### PR TITLE
Bug 580286: Accommodate extra flags with internal builder

### PR DESCRIPTION
**Pull request by jld01**

### CODE REVIEW

**Comment by jld01** [comment]
Original patch from gerrit to re-generate test results.

**Comment by jonahgraham** [comment]
@jld01 - please let me know if you are blocked by anything at all. The only thing I expect is that the `continuous-integration/jenkins/pr-merge` may get marked as failed as any test failures (including flaky tests) causes it to be marked as failure instead of simply equivalent to unstable. I have added a TODO item to #32 to improve this usability issue.

**Comment by jld01** [comment]
> @jld01 - please let me know if you are blocked by anything at all.

Thank you @jonahgraham. The CI jobs seem to be stalled at present, otherwise everything is clear so far.

**Comment by jonahgraham** [comment]
Hmm, I thought it was just a long queue, but it turns out that CI jobs deadlocked. I have cleared the deadlock (I think it relates to the new way I setup to run parallel jobs + the fact that every open PR is rebuilt every time a new commit to main is merged). All that is in flux as I play around and tune settings.

**Comment by jonahgraham** [comment]
One of the test failures looks related to this change: [testDesAddlInVarUserObjs](https://ci.eclipse.org/cdt/job/cdt/job/PR-31-merge/3/testReport/junit/org.eclipse.cdt.managedbuilder.core.tests/BuildDescriptionModelTests/testDesAddlInVarUserObjs/). I am preserving the info available here because I don't know quite how the retention in Jenkins will work yet.

<h1 class="result-failed" style="box-sizing: inherit; line-height: var(--line-height-heading); font-weight: bold; display: block; margin-top: 0.5rem; margin-bottom: 0.5rem; font-size: 2rem; color: rgb(239, 41, 41); font-family: &quot;Libre Franklin&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Failed</h1><p style="box-sizing: inherit; font-family: &quot;Libre Franklin&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 14px; line-height: var(--line-height-base); color: rgb(51, 51, 51); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span style="box-sizing: inherit; font-weight: bold;">org.eclipse.cdt.managedbuilder.core.tests.BuildDescriptionModelTests.testDesAddlInVarUserObjs</span></p><div style="box-sizing: inherit; color: rgb(51, 51, 51); font-family: &quot;Libre Franklin&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; text-align: right;">Failing for the past 3 builds (Since<span> </span><a href="https://ci.eclipse.org/cdt/job/cdt/job/PR-31-merge/1/" class="model-link" style="box-sizing: inherit; text-decoration: none; font-weight: var(--link-font-weight); color: rgb(63, 123, 150); background-color: transparent;"><span class="build-status-icon__wrapper icon-yellow icon-sm" style="box-sizing: inherit; width: 16px; height: 16px; color: var(--unstable-build-icon-color); fill: var(--unstable-build-icon-color); opacity: 1; display: inline-flex; position: relative;"><span class="build-status-icon__outer" style="box-sizing: inherit; display: flex; position: absolute; top: 8px; left: 8px; width: 16px; height: 16px; transform: translate(-50%, -50%);"><svg viewBox="0 0 24 24" focusable="false" class="svg-icon "><use href="/cdt/images/build-status/build-status-sprite.svg#build-status-static"></use></svg></span><svg viewBox="0 0 24 24" focusable="false" class="svg-icon icon-yellow icon-sm"><use href="/cdt/static/3d816110/images/build-status/build-status-sprite.svg#last-unstable"></use></svg></span><span class="jenkins-icon-adjacent" style="box-sizing: inherit; margin-left: 0.5rem; width: 100%;">#1</span></a><span> </span>)</div><div style="box-sizing: inherit; color: rgb(51, 51, 51); font-family: &quot;Libre Franklin&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; text-align: right;"><a href="https://ci.eclipse.org/cdt/job/cdt/job/PR-31-merge/3/testReport/junit/org.eclipse.cdt.managedbuilder.core.tests/BuildDescriptionModelTests/testDesAddlInVarUserObjs/history" style="box-sizing: inherit; text-decoration: none; font-weight: var(--link-font-weight); color: rgb(63, 123, 150); background-color: transparent;">Took 0.49 sec.</a></div><div id="description" style="box-sizing: inherit; color: rgb(51, 51, 51); font-family: &quot;Libre Franklin&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div style="box-sizing: inherit;"></div><div class="jenkins-buttons-row jenkins-buttons-row--invert" style="box-sizing: inherit; display: flex; align-items: center; justify-content: flex-end;"><a class="yui-button link-button" id="description-link" href="https://ci.eclipse.org/cdt/job/cdt/job/PR-31-merge/3/testReport/junit/org.eclipse.cdt.managedbuilder.core.tests/BuildDescriptionModelTests/testDesAddlInVarUserObjs/editDescription" onclick="return replaceDescription();" style="box-sizing: border-box; text-decoration: none; font-weight: var(--btn-link-font-weight); color: var(--primary); background-color: transparent; display: inline-flex; margin-left: 0.25rem; margin-right: 0.25rem; padding: 0.375rem 1rem; min-width: 4.5rem; min-height: 2rem; cursor: pointer; user-select: none; font-size: var(--btn-font-size); line-height: var(--btn-line-height); text-align: center; vertical-align: middle; border: 2px solid transparent; border-radius: 4px; transition: color 0.15s ease-in-out 0s, background-color 0.15s ease-in-out 0s, border-color 0.15s ease-in-out 0s; align-items: center; justify-content: center;"><span class="leading-icon" style="box-sizing: inherit; display: inline-block; margin-right: 0.25rem;"><svg viewBox="0 0 16 16" focusable="false" class="svg-icon icon-small"><use href="/cdt/static/3d816110/images/material-icons/edit.svg#edit"></use></svg></span>Add description</a></div></div>
<h3 style="box-sizing: inherit; line-height: var(--line-height-heading); font-weight: bold; display: block; margin-top: 0.5rem; margin-bottom: 0.5rem; font-size: 1.5rem; color: rgb(51, 51, 51); font-family: &quot;Libre Franklin&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Error Message</h3><pre style="box-sizing: inherit; white-space: pre-wrap; overflow-wrap: break-word; margin: 0px; font-size: 14px; font-family: Consolas, Menlo, Monaco, monospace; background-color: rgb(34, 34, 34); color: rgb(238, 238, 238); padding: 10px 10px 30px; border-radius: 3px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">steps do not match: different number of output types (see console output)</pre><h3 style="box-sizing: inherit; line-height: var(--line-height-heading); font-weight: bold; display: block; margin-top: 0.5rem; margin-bottom: 0.5rem; font-size: 1.5rem; color: rgb(51, 51, 51); font-family: &quot;Libre Franklin&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Stacktrace</h3><pre style="box-sizing: inherit; white-space: pre-wrap; overflow-wrap: break-word; margin: 0px; font-size: 14px; font-family: Consolas, Menlo, Monaco, monospace; background-color: rgb(34, 34, 34); color: rgb(238, 238, 238); padding: 10px 10px 30px; border-radius: 3px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">junit.framework.AssertionFailedError: steps do not match: different number of output types (see console output)
	at org.eclipse.cdt.managedbuilder.core.tests.BuildDescriptionModelTests.doFail(BuildDescriptionModelTests.java:838)
	at org.eclipse.cdt.managedbuilder.core.tests.BuildDescriptionModelTests.doFail(BuildDescriptionModelTests.java:833)
	at org.eclipse.cdt.managedbuilder.core.tests.BuildDescriptionModelTests.stepsMatch(BuildDescriptionModelTests.java:700)
	at org.eclipse.cdt.managedbuilder.core.tests.BuildDescriptionModelTests.stepsMatch(BuildDescriptionModelTests.java:663)
	at org.eclipse.cdt.managedbuilder.core.tests.BuildDescriptionModelTests.doTestStep(BuildDescriptionModelTests.java:584)
	at org.eclipse.cdt.managedbuilder.core.tests.BuildDescriptionModelTests.doTestBuildDescription(BuildDescriptionModelTests.java:865)
	at org.eclipse.cdt.managedbuilder.core.tests.BuildDescriptionModelTests.testDesAddlInVarUserObjs(BuildDescriptionModelTests.java:1781)
</pre><h3 style="box-sizing: inherit; line-height: var(--line-height-heading); font-weight: bold; display: block; margin-top: 0.5rem; margin-bottom: 0.5rem; font-size: 1.5rem; color: rgb(51, 51, 51); font-family: &quot;Libre Franklin&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Standard Output</h3><pre style="box-sizing: inherit; white-space: pre-wrap; overflow-wrap: break-word; margin: 0px; font-size: 14px; font-family: Consolas, Menlo, Monaco, monospace; background-color: rgb(34, 34, 34); color: rgb(238, 238, 238); padding: 10px 10px 30px; border-radius: 3px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">BuildModel[ BuildDescriptionModelTests.testDesAddlInVarUserObjs: steps do not match: different number of output types
Step:
dumping step input step inputs
end dump step
dumping step input step outputs
ioType 0:
/BuildDescription_1/a.c
/BuildDescription_1/b.c
ioType 1:
/BuildDescription_1/c.cpp
/BuildDescription_1/d.cpp
ioType 2:
/BuildDescription_1/e.r
/BuildDescription_1/dir1/g.r
/BuildDescription_1/dir2/h.r
/BuildDescription_1/dir2/i.r
/BuildDescription_1/f.r
end dump step

oStep:
dumping step input step inputs
end dump step
dumping step input step outputs
ioType 0:
/BuildDescription_1/c.cpp
/BuildDescription_1/d.cpp
ioType 1:
/BuildDescription_1/a.c
/BuildDescription_1/b.c
ioType 2:
/BuildDescription_1/e.r
/BuildDescription_1/f.r
/BuildDescription_1/dir1/g.r
/BuildDescription_1/dir2/h.r
/BuildDescription_1/dir2/i.r
ioType 3:
/BuildDescription_1/o1.o
/BuildDescription_1/dir3/o2.o
/BuildDescription_1/dir4/d/o3.o
end dump step
 ]</pre>

**Comment by jld01** [comment]
BuildDescriptionModelTests.testDesAddlInVarUserObjs is now passing and I think the change required to the test makes sense. I would like to undertake some further ad-hoc testing using both external and internal builders with libraries and user objects before merging.

**Comment by jld01** [comment]
I have now tested user objects and libraries for correct command line and correct dependency analysis with both the internal builder and the non-API 'gnu2' makefile generator. I am happy for this to be merged.

**Comment by jonahgraham** [comment]
The test failures now are the flaky tests, unrelated to the change. I’m not really able to review the change myself at the moment.

**Comment by jld01** [comment]
I give @T-Svensson an opportunity for comment, then merge to main, then look at cherry picking to cdt_10_7.

File: build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/buildmodel/BuildDescription.java
Comment (Torbjorn-Svensson) [inline build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/buildmodel/BuildDescription.java:1603]: I would have put `!toolHasExtraFlags` first here since it's cheaper to do that check than the rest.

File: build/org.eclipse.cdt.managedbuilder.core.tests/tests/org/eclipse/cdt/managedbuilder/core/tests/BuildDescriptionModelTests.java
Comment (Torbjorn-Svensson) [inline build/org.eclipse.cdt.managedbuilder.core.tests/tests/org/eclipse/cdt/managedbuilder/core/tests/BuildDescriptionModelTests.java:1759]: Is there no need to check that these files gets handled somewhere else or is it enough to just drop the check?

File: build/org.eclipse.cdt.managedbuilder.core.tests/tests/org/eclipse/cdt/managedbuilder/core/tests/BuildDescriptionModelTests.java
Comment (Torbjorn-Svensson) [inline build/org.eclipse.cdt.managedbuilder.core.tests/tests/org/eclipse/cdt/managedbuilder/core/tests/BuildDescriptionModelTests.java:1656]: Is there no need to check that these files gets handled somewhere else or is it enough to just drop the check?

**Review by Torbjorn-Svensson — APPROVED** [review]
Just some minor comments/questions.
Since there is nothing blocking the change, feel free to submit if you think it's okay as-is.

File: build/org.eclipse.cdt.managedbuilder.core.tests/tests/org/eclipse/cdt/managedbuilder/core/tests/BuildDescriptionModelTests.java
Comment (jld01) [reply build/org.eclipse.cdt.managedbuilder.core.tests/tests/org/eclipse/cdt/managedbuilder/core/tests/BuildDescriptionModelTests.java:1656]: There is nothing further to check in the context of this test since the user objects are no-longer part of the ${INPUTS} of the linker.

File: build/org.eclipse.cdt.managedbuilder.core.tests/tests/org/eclipse/cdt/managedbuilder/core/tests/BuildDescriptionModelTests.java
Comment (jld01) [reply build/org.eclipse.cdt.managedbuilder.core.tests/tests/org/eclipse/cdt/managedbuilder/core/tests/BuildDescriptionModelTests.java:1759]: There is nothing further to check in the context of this test since the user objects are no-longer part of the ${INPUTS} of the linker.

File: build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/buildmodel/BuildDescription.java
Comment (jld01) [reply build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/buildmodel/BuildDescription.java]: Done.

File: build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/buildmodel/BuildDescription.java
Comment (Torbjorn-Svensson) [inline build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/buildmodel/BuildDescription.java:1603]: Have we started using a new code style that includes redundant parenthesis?
In my book, this should be:
```
if (!toolHasExtraFlags && objs != null && objs.length != 0) {
    ...
}
```
But I might have missed something...

Comment (jld01) [reply build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/buildmodel/BuildDescription.java:1603]: We must have different books. :-)

**Comment by jld01** [comment]
Thanks for the review @Torbjorn-Svensson.

**Pull request merged**
